bazel/closure_grpc_web_library.bzl

```diff
@@ -0,0 +1,195 @@
+  root = proto.root.path
+  ws = proto.owner.workspace_root
+  if path.startswith(root):
+    path = path[len(root):]
+  if path.startswith("/"):
+    path = path[1:]
+  if path.startswith(ws):
+    path = path[len(ws):]
+  if path.startswith("/"):
+    path = path[1:]
+  return path
+
+def _proto_include_path(proto):
+  path = proto.path[:-len(_proto_path(proto))]
+  if not path:
+    return '.'
+  if path.endswith("/"):
+    path = path[:-1]
+  return path
+
+def _proto_include_paths(protos):
+  return depset([_proto_include_path(proto) for proto in protos])
+
+def _generate_closure_grpc_web_src_progress_message(name):
+  # TODO(yannic): Add a better message?
+  return 'Generating GRPC Web %s' % name
+
+def _generate_closure_grpc_web_srcs(
+    actions, protoc, protoc_gen_grpc_web, import_style, mode,
+    sources, transitive_sources):
+  all_sources = [src for src in sources] + [src for src in transitive_sources]
+  proto_include_paths = [
+      '-I%s' % p for p in _proto_include_paths(
+          [f for f in all_sources])
+  ]
+
+  grpc_web_out_common_options = ','.join([
+      'import_style={}'.format(import_style),
+      'mode={}'.format(mode),
+  ])
+
+  files = []
+  for src in sources:
+    name = '{}.grpc.js'.format(
+        '.'.join(src.path.split('/')[-1].split('.')[:-1]))
+    js = actions.declare_file(name)
+    files.append(js)
+
+    args = proto_include_paths + [
+        '--plugin=protoc-gen-grpc-web={}'.format(protoc_gen_grpc_web.path),
+        '--grpc-web_out={options},out={out_file}:{path}'.format(
+            options = grpc_web_out_common_options,
+            out_file = name,
+            path = js.path[:js.path.rfind('/')],
+        ),
+        src.path
+    ]
+
+    actions.run(
+        inputs = [protoc_gen_grpc_web] + all_sources,
+        outputs = [js],
+        executable = protoc,
+        arguments = args,
+        progress_message =
+            _generate_closure_grpc_web_src_progress_message(name),
+    )
+
+  return files
+
+_error_multiple_deps = ''.join([
+  "'deps' attribute must contain exactly one label ",
+  "(we didn't name it 'dep' for consistency). ",
+  "We may revisit this restriction later.",
+])
+
+def _closure_grpc_web_library_impl(ctx):
+  if len(ctx.attr.deps) > 1:
+    # TODO(yannic): Revisit this restriction.
+    fail(_error_multiple_deps, 'deps');
+
+  dep = ctx.attr.deps[0]
+
+  srcs = _generate_closure_grpc_web_srcs(
+      actions = ctx.actions,
+      protoc = ctx.executable._protoc,
+      protoc_gen_grpc_web = ctx.executable._protoc_gen_grpc_web,
+      import_style = 'closure',
+      mode = 'grpcwebtext',
+      sources = dep.proto.direct_sources,
+      transitive_sources = dep.proto.transitive_imports,
+  )
+
+  deps = unfurl(ctx.attr.deps, provider = "closure_js_library")
+  deps += [
+      ctx.attr._grpc_web_abstractclientbase,
+      ctx.attr._grpc_web_clientreadablestream,
+      ctx.attr._grpc_web_error,
+      ctx.attr._grpc_web_grpcwebclientbase
+  ]
```

Comment:
I wonder why only these 4 classes? Why not `Status` and `StatusCode` etc?

Reply:
We only need to add the directly `goog.required` classes to deps, `Status`,  `StatusCode`, etc are included as transitive deps by those 4 classes.
https://github.com/grpc/grpc-web/blob/master/javascript/net/grpc/web/grpc_generator.cc#L409
(Only mode `GrpcWeb` seems to be open source).